### PR TITLE
Add a clear documentation for this repo and ExecWrappedStudentCodeTest

### PR DESCRIPTION
**Pull request by manhhung741**
I'm making use of `xqueue_watcher` for a custom deployment of the Open Edx system, and I found this repo has a lot of good stuff. However, the lack of documentation made it very hard to use. I spent a lot of time testing and reading code to understand how to use it. Here is documentation of my understanding for posterity. If there is any mistake, please help me correct my understanding. I appreciate it! 

And thanks for the code. It mostly works out of the box!

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @manhhung741! I've created [OSPR-5355](https://openedx.atlassian.net/browse/OSPR-5355) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.


:warning: We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](https://open.edx.org/wp-content/uploads/2019/01/individual-contributor-agreement.pdf)     or indicated your institutional affiliation. Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

**Comment by natabene** [comment]
@manhhung741 Thank you for the contribution! When you have a chance, could you please send in a signed Contributor Agreement mentioned in the earlier comment? You only need to do this once and we can then review all your future code contributions. Thanks

**Comment by manhhung741** [comment]
@natabene Thanks! I've sent the signed ICA to legal@edx.org!

**Comment by natabene** [comment]
@manhhung741 Thank you for joining the Open edX community! We have added your user to our records and can now review your code going forward.

File: README.md [hunk 1]
```diff
@@ -3,58 +3,153 @@ xqueue_watcher
...
+   │   └── answer.py  # - if using JailedGrader
+   ├── ...
+   └── exercise2/
+       ├── grader.py
+       └── answer.py
+```
+Running XQueue Watcher:
+======================
+
+Usually you can run XQueue Watcher without making any changes. You should keep course-specific in another folder like shown above, so that you can update xqueue_watcher anytime.
```
Comment (sarina) [inline README.md:36]: `you should keep course-specific in another folder` --> course-specific what? Files, I presume - good to add a noun here :)

File: README.md [hunk 2]
```diff
@@ -3,58 +3,153 @@ xqueue_watcher
...
+python -m xqueue_watcher -d [path to the config directory, eg ../config]
+```
+
+The course configuration JSON file in `conf.d` should have the following structure:
+```json
+    {
+        "test-123": {
+            "SERVER": "http://127.0.0.1:18040",
+            "CONNECTIONS": 1,
+            "AUTH": ["lms", "lms"],
```
Comment (sarina) [inline README.md:55]: Since this is a username/password pair I recommend being more explicit

`"AUTH": ["uname", "pwd"],`

File: README.md [hunk 3]
```diff
@@ -3,58 +3,153 @@ xqueue_watcher
...
 * `AUTH`: list of username, password
 * `CONNECTIONS`: how many threads to spawn to watch the queue
 * `HANDLERS`: list of callables that will be called for each queue submission
-	* `HANDLER`: callable name
-	* `KWARGS`: optional keyword arguments to apply during instantiation
+   * `HANDLER`: callable name, see below for Submissions Handler
+   * `KWARGS`: optional keyword arguments to apply during instantiation
+      * `grader_root`: path to the course directory, eg /path/to/my-course
 
+> TODO: document logging.json
```
Comment (sarina) [inline README.md:77]: Is this something you're able to document now? 😄

File: README.md [hunk 4]
```diff
@@ -3,58 +3,153 @@ xqueue_watcher
...
 
-xqueue_watcher.grader.Grader
-========================
+Submissions Handler
+===================
+
+When xqueue_watcher detects any new submission, it will be passed to the submission handler for grading. It will instantiate a new handler based on the name configured above, with submission information retrieved
```
Comment (sarina) [inline README.md:82]: I recommend making xqueue_watcher into code format (`xqueue_watcher`)

File: README.md [hunk 5]
```diff
@@ -3,58 +3,153 @@ xqueue_watcher
...
+from XQueue. There is a base grader defined in xqueue_watcher: Grader and JailedGrader (for Python, using CodeJail). If you don't use JailedGrader, you'd have to implement your own Grader by subclassing `xqueue_watcher.grader.Grader
```
Comment (sarina) [inline README.md:83]: I recommend making xqueue_watcher into code format (`xqueue_watcher`) along with Grader, JailedGrader, CodeJail.

You're also missing a closing backtick on `xqueue_watcher.grader.Grader`

File: README.md [hunk 6]
```diff
@@ -3,58 +3,153 @@ xqueue_watcher
...
+
+The payload from XQueue will be a JSON that usually looks like this, notice that "grader" is a required field in the "grader_payload" and must be configured accordingly in the Studio for the exercise.
```
Comment (sarina) [inline README.md:85]: Could you briefly explain how to configure in Studio?

File: README.md [hunk 7]
```diff
@@ -3,58 +3,153 @@ xqueue_watcher
...
-Subclass xqueue_watcher.grader.Grader and override the `grade` method. Then add your grader to the config like `"handler": "my_module.MyGrader"`. The arguments for the `grade` method are:
-	* `grader_path`: absolute path to the grader defined for the current problem
-	* `grader_config`: other configuration particular to the problem
-	* `student_response`: student-supplied code
+Subclass `xqueue_watcher.grader.Grader` and override the `grade` method. Then add your grader to the config like `"handler": "my_module.MyGrader"`. The arguments for the `grade` method are:
+   * `grader_path`: absolute path to the grader defined for the current problem.
+   * `grader_config`: other configuration particular to the problem
+   * `student_response`: student-supplied code
 
+Note that `grader_path` is constructed by appending the relative path to the grader from `grader_payload` to the `grader_root` in the configuration JSON. If the handler cannot find a `grader.py` file, it would fail to grade the submission.
```
Comment (sarina) [inline README.md:106]: Does it fail in any obvious way that could be called out?

File: README.md [hunk 8]
```diff
@@ -3,58 +3,153 @@ xqueue_watcher
...
+To sandbox python, use [CodeJail](https://github.com/edx/codejail). In your handler configuration, add:
+```json
+    "HANDLER": "xqueue_watcher.jailedgrader.JailedGrader",
+    "CODEJAIL": {
+        "name": "python",
+        "python_bin": "/path/to/sandbox/python",
+        "user": "sandbox_username"
+    }
+```
 Then, `codejail_python` will automatically be added to the kwargs for your handler. You can then import codejail.jail_code and run `jail_code("python", code...)`. You can define multiple sandboxes and use them as in `jail_code("special-python", ...)`
```
Comment (sarina) [inline README.md:122]: what does `special-python` mean? This last sentence confuses me.

File: README.md [hunk 9]
```diff
@@ -3,58 +3,153 @@ xqueue_watcher
...
+import gradelib
+
+grader = gradelib.Grader()
+
+# execute a raw student code & capture stdout
+grader.add_preprocessor(gradelib.wrap_in_string)
+grader.add_test(gradelib.ExecWrappedStudentCodeTest({}, "basic test"))
+```
+
+You can also write your own test class, processor and input checks.
```
Comment (sarina) [inline README.md:155]: I would love to see an example of an actual grading file, if you have one!

**Review by sarina — CHANGES_REQUESTED** [review]
Hiya! This is really excellent to have documentation in this repo. I have left some feedback comments I'd like to see addressed. Please ping me if you have any questions and I'll answer promptly.

File: README.md [hunk 10]
```diff
@@ -3,58 +3,153 @@ xqueue_watcher
...
+                    }
+                }
+            ]
+        }
+    }
+```
 
 * `test-123`: the name of the queue
 * `SERVER`: XQueue server address
 * `AUTH`: list of username, password
```
Comment (doctoryes) [inline README.md:70]: Nit: It's be nice to be clear for `AUTH` that it's the username/password for a XQueue Django user. Maybe:
`* 'AUTH': List containing [username, password] of XQueue Django user`
I really wish that `AUTH` was more descriptive - but that's beyond this PR scope.

File: README.md (hunk 3 above)
Comment (doctoryes) [reply README.md:77]: This TODO should either be 'to-done' or removed. :)

File: README.md (hunk 5 above)
Comment (doctoryes) [reply README.md:83]: Nit: `There is a base grader defined...` -> `Base graders are defined...`

File: README.md (hunk 6 above)
Comment (doctoryes) [reply README.md:85]: Nit: Maybe instead:
`The payload received from XQueue will be a JSON object that usually looks like the JSON below. Note that "grader" is a required field in the "grader_payload" and must be configured accordingly in Studio.`?

File: README.md (hunk 8 above)
Comment (doctoryes) [reply README.md:122]: Maybe `special-python` -> `python-version1`/`another-python-version`?

File: README.md [hunk 11]
```diff
@@ -3,58 +3,153 @@ xqueue_watcher
...
+    "HANDLER": "xqueue_watcher.jailedgrader.JailedGrader",
+    "CODEJAIL": {
+        "name": "python",
+        "python_bin": "/path/to/sandbox/python",
+        "user": "sandbox_username"
+    }
+```
 Then, `codejail_python` will automatically be added to the kwargs for your handler. You can then import codejail.jail_code and run `jail_code("python", code...)`. You can define multiple sandboxes and use them as in `jail_code("special-python", ...)`
+
+To use JailedGrader, you also need to provide an `answer.py` file on the same folder with the `grader.py` file. The grader will run both student submission and `answer.py` and compare the output with each other.
```
Comment (doctoryes) [inline README.md:124]: Could you please mention that `answer.py` contains an actual correct answer to the problem/exercise?

File: README.md [hunk 12]
```diff
@@ -3,58 +3,153 @@ xqueue_watcher
...
+
+### Grading Support utilities
+There are several grading support utilities that make writing `grader.py` for python code easy. Check out
+`grader_support/gradelib.py` for the documentation.
+
+- `grader_support.gradelib.Grader`: a base class for creating a new submission grader. Not to be confused with `xqueue-watcher.grader.Grader`. You can add input checks, preprocessors and tests to a grader object.
```
Comment (doctoryes) [inline README.md:130]: Nit: `...to a grader object` -> `to a Grader object`.

File: grader_support/gradelib.py [hunk 13]
```diff
@@ -546,6 +546,17 @@ def __init__(self, fn_name, args, environment=None, output_writer=None, short_de
             short_desc = "Test: %s(%s)" % (fn_name, ", ".join(repr(a) for a in args))
         Test.__init__(self, test_fn, short_desc, detailed_desc, compare)
 
+class ExecWrappedStudentCodeTest(Test):
+    """
+    A Test that exec student code and capture the stdout result.
```
Comment (doctoryes) [inline grader_support/gradelib.py:551]: Nit: `Test that executes the student code and captures the result, which is output to stdout.`

**Review by doctoryes — COMMENTED** [review]
Thanks for adding more documentation to this repo!

**Comment by natabene** [comment]
@sarina @doctoryes Glad to here there is still a chance for this PR! Thanks for hopefully taking it over it finish line soon.

**Comment by sarina** [comment]
@manhhung741 is there any chance you are willing to follow up on this pull request? If you are not able to, we will try to get your changes in ourselves. Could you please let us know by the end of this week (23 May) if you could finish this work? Thank you 😄

**Comment by manhhung741** [comment]
Sorry I haven't had time to revise the PR. If you would like to take over it, please feel free!

**Comment by manhhung741** [comment]
@sarina ^ 
Again, sorry for the late reply.

**Comment by sarina** [comment]
I will try to make as many changes as possible over in https://github.com/edx/xqueue-watcher/pull/66 and ping Julia when ready for review.

**Comment by edx-webhook** [comment]
@manhhung741 Even though your pull request wasn’t merged, please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/xqueue-watcher&entry.867055334=https://github.com/edx/xqueue-watcher/pull/54&entry.1484655318=https://github.com/manhhung741&entry.752974735=2021-01-12+07:21&entry.1917517419=2021-07-30+15:32&entry.2133058324=No) so we can improve your experience in the future.

**Pull request closed without merging**